OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -56,46 +65,86 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
 				activeResupplyTypes |= ResupplyType.Rearm;
 		}
 
-		protected override void OnFirstRun(Actor self)
-		{
-			if (host.Type == TargetType.Invalid)
-				return;
-
-			if (activeResupplyTypes > 0)
-				foreach (var notifyResupply in notifyResupplies)
-					notifyResupply.BeforeResupply(host.Actor, self, activeResupplyTypes);
-
-			// Reset the ReloadDelay to avoid any issues with early cancellation
-			// from previous reload attempts (explicit order, host building died, etc).
-			// HACK: this really shouldn't be managed from here
-			if (activeResupplyTypes.HasFlag(ResupplyType.Rearm))
-				foreach (var pool in rearmable.RearmableAmmoPools)
-					pool.RemainingTicks = pool.Info.ReloadDelay;
-		}
-
 		public override bool Tick(Actor self)
 		{
 			// HACK: If the activity is cancelled while we're already resupplying (or about to start resupplying),
-			// move actor outside the resupplier footprint
-			// TODO: This check is nowhere near robust enough, and should be rewritten
+			// move actor outside the resupplier footprint.
+			// TODO: This check is nowhere near robust enough, and should be rewritten.
 			if (IsCanceling && host.IsInRange(self.CenterPosition, closeEnough))
 			{
-				QueueChild(self.Trait<IMove>().MoveToTarget(self, host));
 				foreach (var notifyResupply in notifyResupplies)
 					notifyResupply.ResupplyTick(host.Actor, self, ResupplyType.None);
 
-				return false;
-			}
+				if (aircraft != null)
+				{
+					aircraft.AllowYieldingReservation();
+					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
+						Queue(new TakeOff(self));
```

Comment:
`ReturnToBase` already takes care of queueing the `TakeOff` so this seems a bit redundant. Also, is there a reason why this is not queued as a childactivity? This wouldn't work correctly when other activities are queued after `Resupply.`

Reply:
I've lost track of the specifics about `ReturnToBase` and taking off, but having the `TakeOff` `Queue`d after other user-ordered activities is definitely an issue.